packages/core/src/entrypoint/index.ts

```diff
@@ -0,0 +1,190 @@
+import { type Chain } from "viem";
+import { EntryPointAbi_v6 } from "../abis/EntryPointAbi_v6.js";
+import { EntryPointAbi_v7 } from "../abis/EntryPointAbi_v7.js";
+import {
+  EntryPointNotFoundError,
+  MismatchingEntryPointError,
+} from "../errors/entrypoint.js";
+import type { UserOperationLike } from "../types.js";
+import EntryPoint_v6 from "./0.6.js";
+import EntryPoint_v7 from "./0.7.js";
+import type {
+  DefaultEntryPointVersion,
+  EntryPointDef,
+  EntryPointDefRegistry,
+  EntryPointRegistry,
+  EntryPointVersion,
+  GetEntryPointOptions,
+  SupportedEntryPoint,
+} from "./types.js";
+
+export const defaultEntryPointVersion: DefaultEntryPointVersion = "0.6.0";
+
+export const entryPointRegistry: EntryPointRegistry = {
+  "0.6.0": EntryPoint_v6,
+  "0.7.0": EntryPoint_v7,
+};
+
+export const coerceEntryPointVersion = <
+  TEntryPointVersion extends EntryPointVersion
+>(
+  entryPoint: EntryPointRegistry[EntryPointVersion],
+  userOperation: UserOperationLike,
+  throwable?: boolean
+): UserOperationLike<TEntryPointVersion> | undefined => {
+  if (entryPoint.isUserOpVersion(userOperation)) {
+    return userOperation as UserOperationLike<TEntryPointVersion>;
+  }
+  if (!throwable) {
+    return undefined;
+  }
+  throw new MismatchingEntryPointError(entryPoint.version, userOperation);
+};
+
+export function getEntryPoint<
+  TEntryPointVersion extends EntryPointVersion = DefaultEntryPointVersion,
+  TChain extends Chain = Chain
+>(
+  chain: TChain,
+  options: GetEntryPointOptions<TEntryPointVersion>
+): EntryPointDefRegistry<TChain>[TEntryPointVersion];
+
+export function getEntryPoint<
+  TEntryPointVersion extends DefaultEntryPointVersion = DefaultEntryPointVersion,
+  TChain extends Chain = Chain
+>(
+  chain: TChain,
+  options?: GetEntryPointOptions<TEntryPointVersion>
+): EntryPointDefRegistry<TChain>[TEntryPointVersion];
+
+export function getEntryPoint<TChain extends Chain = Chain>(
+  chain: TChain,
+  options?: GetEntryPointOptions<DefaultEntryPointVersion>
+): EntryPointDefRegistry<TChain>[DefaultEntryPointVersion];
+
+export function getEntryPoint<
+  TEntryPointVersion extends EntryPointVersion,
+  TChain extends Chain = Chain
+>(
+  chain: TChain,
+  options: GetEntryPointOptions<TEntryPointVersion>
+): EntryPointDefRegistry<TChain>[EntryPointVersion] {
+  const { version = defaultEntryPointVersion, addressOverride } = options ?? {
+    version: defaultEntryPointVersion,
+  };
+
+  const entryPoint = entryPointRegistry[version ?? defaultEntryPointVersion];
+  const address = addressOverride || entryPoint.address[chain.id];
+  if (!address) {
+    throw new EntryPointNotFoundError(chain, version);
+  }
+
+  if (version === "0.6.0") {
+    const v6 = entryPoint as SupportedEntryPoint<
+      "0.6.0",
+      TChain,
+      typeof EntryPointAbi_v6
+    >;
```

Comment:
you can avoid the typecast and need for the `v6` const completely here by just checking:
```
if (entryPoint.version === "0.6.0")
```

This will then allow typescript to infer the the type of `entryPoint` is already the v6 one and same below

Reply:
no, this one is not a entrypoint version generic function. just checking for version === '0.6.0' does not return the type inferred result.

Reply:
sorry I'm not saying use `version === 0.6.0` but `entryPoint.version === 0.6.0` that will be inferred correctly

Reply:
not only does it work, but if you do that you don't even need the typecasting. 

Typecasting is generally a code smell. if things are typed correctly and the logic adheres by to the types, then you shouldn't need to do any of that.

because it's easier than making the suggestions in github this is what I mean: https://github.com/alchemyplatform/aa-sdk/pull/549